scio/src/main/scala/org/apache/zeppelin/scio/ScioInterpreter.scala

```diff
@@ -0,0 +1,164 @@
+    // There should be no harm if we keep this for sbt launch.
+    val thisJar = this.getClass.getProtectionDomain.getCodeSource.getLocation.getPath
+    // In some cases this may be `target/classes`
+    if(thisJar.endsWith(".jar")) {
+      settings.plugin.appendToValue(thisJar)
+    }
+
+    ClassPath.split(settings.classpath.value)
+      .find(File(_).name.startsWith("paradise_"))
+      .foreach(settings.plugin.appendToValue)
+
+    // Force the repl to be synchronous, so all cmds are executed in the same thread
+    settings.Yreplsync.value = true
+
+    val scioClassLoader = new ScioReplClassLoader(
+      ClassPath.toURLs(settings.classpath.value).toArray ++
+        classLoaderURLs(Thread.currentThread().getContextClassLoader),
+      null,
+      Thread.currentThread.getContextClassLoader)
+
+    REPL = new ScioILoop(scioClassLoader, args, None, new JPrintWriter(innerOut))
+    scioClassLoader.setRepl(REPL)
+
+    // Set classloader chain - expose top level abstract class loader down
+    // the chain to allow for readObject and latestUserDefinedLoader
+    // See https://gist.github.com/harrah/404272
+    settings.embeddedDefaults(scioClassLoader)
+
+    // No need for bigquery dumps
+    sys.props("bigquery.plugin.disable.dump") = "true"
+
+    REPL.settings_=(settings)
+    REPL.createInterpreter()
+  }
+
+  override def close(): Unit = {
+    logger.info("Closing Scio interpreter!")
+    REPL.closeInterpreter()
+  }
+
+  override def interpret(code: String, context: InterpreterContext): InterpreterResult = {
+    val paragraphId = context.getParagraphId
+
+    val consoleOut = new PrintStream(innerOut)
+    System.setOut(consoleOut)
+    innerOut.setInterpreterOutput(context.out)
+
+    try {
+      import tools.nsc.interpreter.Results._
+      REPL.interpret(code) match {
+        case Success => {
+          logger.debug(s"Successfully executed `$code` in $paragraphId")
+          new InterpreterResult(InterpreterResult.Code.SUCCESS)
+        }
+        case Error => {
+          logger.error(s"Error executing `$code` in $paragraphId")
+          new InterpreterResult(InterpreterResult.Code.ERROR)
+        }
+        case Incomplete => {
+          logger.warn(s"Code `$code` not complete in $paragraphId")
+          new InterpreterResult(InterpreterResult.Code.INCOMPLETE, "Incomplete expression")
+        }
+      }
+    } catch {
+      case e: Exception =>
+        logger.info("Interpreter exception", e)
+        new InterpreterResult(InterpreterResult.Code.ERROR, e.getMessage)
+    } finally {
+      innerOut.flush()
+      innerOut.setInterpreterOutput(null)
+      consoleOut.flush()
+    }
+  }
+
+  override def cancel(context: InterpreterContext): Unit = {
+    // not implemented
+  }
+
+  override def getFormType: FormType = {
+    FormType.NATIVE
+  }
+
+  override def getProgress(context: InterpreterContext): Int = {
+    // not implemented
+    42
+  }
```

Comment:
The convention is 0 if progress is not implemented IIRC. If the execution is started, the state would transite from `PENDING` to `RUNNING`